packages/scm/src/browser/scm-tree-widget.tsx

```diff
@@ -0,0 +1,990 @@
+
+    @inject(MenuModelRegistry) protected readonly menus: MenuModelRegistry;
+    @inject(CommandRegistry) protected readonly commands: CommandRegistry;
+    @inject(CorePreferences) protected readonly corePreferences: CorePreferences;
+    @inject(ScmContextKeyService) protected readonly contextKeys: ScmContextKeyService;
+    @inject(LabelProvider) protected readonly labelProvider: LabelProvider;
+    @inject(EditorManager) protected readonly editorManager: EditorManager;
+    @inject(DiffNavigatorProvider) protected readonly diffNavigatorProvider: DiffNavigatorProvider;
+    @inject(IconThemeService) protected readonly iconThemeService: IconThemeService;
+
+    protected readonly toDisposeOnRepositoryChange = new DisposableCollection();
+
+    constructor(
+        @inject(TreeProps) readonly props: ScmTreeProps,
+        @inject(TreeModel) readonly model: TreeModel,
+        @inject(ContextMenuRenderer) protected readonly contextMenuRenderer: ContextMenuRenderer,
+        @inject(ScmService) protected readonly scmService: ScmService,
+    ) {
+        super(props, model, contextMenuRenderer);
+        this.id = 'resource_widget';
+    }
+
+    @postConstruct()
+    protected init(): void {
+        super.init();
+        this.addClass('groups-outer-container');
+
+        this.refreshOnRepositoryChange();
+        this.toDispose.push(this.scmService.onDidChangeSelectedRepository(() => {
+            this.refreshOnRepositoryChange();
+            this.forceUpdate();
+        }));
+        this.toDispose.push(this.labelProvider.onDidChange(event => {
+            this.update();
+        }));
+    }
+
+    protected createTree(repository: ScmRepository): FileChangeGroupRoot {
+        const root = {
+            id: 'file-change-tree-root',
+            parent: undefined,
+            visible: false,
+            rootUri: repository.provider.rootUri,
+            children: []
+        } as FileChangeGroupRoot;
+
+        const { groups } = repository.provider;
+        const groupNodes = groups
+            .filter(group => !!group.resources.length || !group.hideWhenEmpty)
+            .map(group => this.toGroupNode(group, root));
+        root.children = groupNodes;
+
+        return root;
+    }
+
+    protected _viewMode: 'tree' | 'flat' = 'flat';
+    set viewMode(id: 'tree' | 'flat') {
+        const oldSelection = this.model.selectedNodes;
+        this._viewMode = id;
+        const repository = this.scmService.selectedRepository;
+        if (repository) {
+            this.model.root = this.createTree(repository);
+
+            for (const oldSelectedNode of oldSelection) {
+                const newNode = this.model.getNode(oldSelectedNode.id);
+                if (SelectableTreeNode.is(newNode)) {
+                    this.revealNode(newNode);  // this call can run asynchronously
+                }
+            }
+        }
+    }
+    get viewMode(): 'tree' | 'flat' {
+        return this._viewMode;
```

Comment:
It should be rather `widget.model.mode` to have clear separation of rendering and behaviour

Reply:
Now moved to the new model.